runtime/bin/mach_o_loader.cc

```diff
@@ -0,0 +1,466 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+#include "bin/mach_o_loader.h"
+#include <sys/types.h>
+
+#include "platform/globals.h"
+
+#include <cstdio>
+#include <functional>
+#include <memory>
+#include <utility>
+
+#include "bin/file.h"
+#include "bin/virtual_memory.h"
+#include "platform/utils.h"
+
+#include "platform/mach_o.h"
+#include "platform/syslog.h"
+#include "platform/unwinding_records.h"
+#include "vm/compiler/runtime_api.h"
+
+namespace dart {
+namespace bin {
+
+namespace mach_o {
+
+// This mappable abstraction matches what ElfLoader uses, but probably
+// could be removed since we don't need to support loading from memory.
+class Mappable {
```

Comment:
I tried to model this off of what elf_loader does for easiest maintenance.  I did not try to share code between them, but someone certainly could do that in a follow-up pass.  Both of them likely need to get moved into the VM instead of bin/